kedro/contrib/io/cached/cached.py

```diff
@@ -0,0 +1,121 @@
+"""
+import pickle
+from typing import Any, Dict
+
+from kedro.io import (
+    AbstractDataSet,
+    ExistsMixin,
+    FilepathVersionMixIn,
+    MemoryDataSet,
+    Version,
+)
+from kedro.io.core import VERSIONED_FLAG_KEY
+
+
+class CachedDataSet(AbstractDataSet, ExistsMixin, FilepathVersionMixIn):
+    """
+    ``CachedDataSet`` is a dataset wrapper which caches in memory the data saved, so that the
+    user avoids io operations with slow storage media.
+
+    You can also specify ``CachedDataSet``s in catalog.yml as follows:
+    ::
+      test_ds:
+          type: kedro.contrib.io.cached.CachedDataSet
+          versioned: true
+          dataset:
+              type: CSVLocalDataSet
+              filepath: example.csv
+
+    Please notice that if your dataset is versioned, this should be indicated in the wrapper class
+    as shown above
+    """
+
+    def __init__(
+        self, dataset: AbstractDataSet, max_loads: int = None, version: Version = None
+    ):
+        super().__init__()
+        if isinstance(dataset, Dict):
+            self._dataset = self._from_config(dataset, version or Version(None, None))
+        elif isinstance(dataset, AbstractDataSet):
+            self._dataset = dataset
+        else:
+            raise ValueError(
+                "The argument type of `dataset` should be either a dict/YML "
+                "representation of the dataset, or the actual dataset object."
+            )
+        self._cache = MemoryDataSet(max_loads=max_loads)
+
+    @staticmethod
+    def _from_config(config, version):
+        if VERSIONED_FLAG_KEY in config:
+            raise ValueError(
+                "Cached datasets should specify that they are versioned in the "
+                "`CachedDataSet`, not in the wrapped dataset."
+            )
+        if version:
+            config[VERSIONED_FLAG_KEY] = version
+        return AbstractDataSet.from_config("Cached", config, version.load, version.save)
+
+    def _describe(self) -> Dict[str, Any]:
+        return {
+            "dataset": self._dataset._describe(),  # pylint: disable=protected-access
+            "cache": self._cache._describe(),  # pylint: disable=protected-access
+        }
+
+    def _load(self):
+        if self._cache.exists():
+            return self._cache.load()
+
+        data = self._dataset.load()
+        self._cache.save(data)
+        return data
+
+    def _save(self, data: Any) -> None:
+        self._dataset.save(data)
+        self._cache.save(data)
+
+    def _exists(self) -> bool:
+        def _wrapped_exists():
+            if hasattr(self._dataset, "exists"):
+                return self._dataset.exists()
+            return False  # pragma: no cover
+
+        return self._cache.exists() or _wrapped_exists()
+
+    def __getstate__(self):
+        # in the future this can be made pickleable by erasing the cache.
+        # A slight challenge would be to see how to propagate max_loads
+        raise pickle.PicklingError(
```

Comment:
if you used managed memory datasets like the parallel runner then this would be parallel runner compatible, and I'd guess anyone who cares about caching datasets is also going to want parallel
it'd be nice if had some way to indicate that a thing can be used with managed memory, but is otherwise not generally serializable, at the moment airflow explicitly checks for memory datasets which isn't great and wouldn't catch this

Reply:
On one hand I agree, on the other hand if you have to serialise to move to a separate process caching kind of defeats the purpose here tho, no? Although I agree that this can be much faster for slow formats or network datasets.

> at the moment airflow explicitly checks for memory datasets which isn't great and wouldn't catch this

Maybe inherit from `MemoryDataSet`? But this feels wrong as it conveys a completely wrong message..

Shall I take a stab converting this to a managed memory dataset, or would that be an overkill? What I think I should probably do is just drop the cache when serialising..

Reply:
I'd leave it as is and address it in a followup